master/pkg/model/task.go

```diff
@@ -148,57 +148,31 @@ func MostProgressedAllocationState(states ...AllocationState) AllocationState {
 		return AllocationStatePending
 	}
 
-	max := states[0]
-	for _, state := range states {
-		if state > max {
-			max = state
-		}
+	// Can't use taskv1.State_value[state] since in proto
+	// "STATE_TERMINATING" > "STATE_TERMINATED"
+	// while our model used to have
+	// "STATE_TERMINATED" > "STATE_TERMINATING".
+	statesToOrder := map[AllocationState]int{
+		AllocationStatePending:     0,
+		AllocationStateAssigned:    1,
+		AllocationStatePulling:     2,
+		AllocationStateStarting:    3,
+		AllocationStateRunning:     4,
+		AllocationStateTerminating: 5,
+		AllocationStateTerminated:  6,
 	}
-	return max
-}
-
-// String returns the string representation of the task state.
-func (s AllocationState) String() string {
-	switch s {
-	case AllocationStatePending:
-		return "PENDING"
-	case AllocationStateAssigned:
-		return "ASSIGNED"
-	case AllocationStatePulling:
-		return "PULLING"
-	case AllocationStateStarting:
-		return "STARTING"
-	case AllocationStateRunning:
-		return "RUNNING"
-	case AllocationStateTerminating:
-		return "TERMINATING"
-	case AllocationStateTerminated:
-		return "TERMINATED"
-	default:
-		return "UNSPECIFIED"
+	maxOrder, state := statesToOrder[states[0]], states[0]
+	for _, s := range states {
+		if order := statesToOrder[s]; order > maxOrder {
+			maxOrder, state = order, s
+		}
 	}
+	return state
 }
 
 // Proto returns the proto representation of the task state.
 func (s AllocationState) Proto() taskv1.State {
-	switch s {
-	case AllocationStatePending:
-		return taskv1.State_STATE_PENDING
-	case AllocationStateAssigned:
-		return taskv1.State_STATE_ASSIGNED
-	case AllocationStatePulling:
-		return taskv1.State_STATE_PULLING
-	case AllocationStateStarting:
-		return taskv1.State_STATE_STARTING
-	case AllocationStateRunning:
-		return taskv1.State_STATE_RUNNING
-	case AllocationStateTerminating:
-		return taskv1.State_STATE_TERMINATING
-	case AllocationStateTerminated:
-		return taskv1.State_STATE_TERMINATED
-	default:
-		return taskv1.State_STATE_UNSPECIFIED
-	}
+	return taskv1.State(taskv1.State_value[fmt.Sprintf("STATE_%s", s)])
```

Comment:
idk, I still prefer the exhaustive switch here.. or if there was a test to ensure proto and go (like the go and pg test) were always in sync this is fine.

Reply:
Changed it back. 

Just wondering what is the reasoning behind not just using the Proto enum inside the Go model?